### ISSUE TITLE
[ENH] Permutation-based Causal Discovery (BOSS)

### ISSUE THREAD

**Issue by anusa-saha** [pgmpy/pgmpy#2912]
@ankurankan 

### Description
pgmpy currently supports several structure learning algorithms such as HillClimbSearch, ExhaustiveSearch, GES, MMHC, and PC. However, permutation-based causal discovery algorithms are not currently available.

Permutation-based methods search over variable orderings instead of the entire DAG space, which can make the optimization problem significantly more tractable while still guaranteeing acyclicity.

One promising algorithm in this class is **BOSS (Best Order Score Search)**. It greedily searches over permutations of variables and constructs the best DAG consistent with each ordering using score-based methods such as BIC. The method also introduces **Grow-Shrink Trees (GST)** to efficiently cache intermediate scoring computations.

### Why this is useful
Permutation-based approaches offer several advantages:
- Efficient search compared to full DAG space exploration
- Natural DAG guarantees through ordering constraints
- Competitive performance on high-dimensional datasets
- Complements existing score-based approaches like HillClimbSearch and GES

### Proposed Contribution
I would like to work on implementing a permutation-based causal discovery estimator for pgmpy, starting with **BOSS**.

### Reference
Fast Scalable and Accurate Discovery of DAGs Using the Best Order Score Search and Grow-Shrink Trees
https://arxiv.org/abs/2310.17679

**Comment by anusa-saha** [comment]
@ankurankan working on the API structure

**Comment by anusa-saha** [comment]
The proposed way of implementing this 
The algorithm will be added to: `pgmpy/causal_discovery/BOSS.py`

Since BOSS performs permutation-based causal discovery, it should inherit from: `_BaseCausalDiscovery` which is defined in: `pgmpy/causal_discovery/_base.py`

Public API
`class BOSS(_BaseCausalDiscovery)`
Constructor used:
```
def __init__(
    self,
    scoring_method: Optional[str] = "bic",
    start_perm: Optional[List[str]] = None,
    use_bes: bool = False,
    max_iter: int = 100,
    random_state: Optional[int] = None,
):
```
| Parameter        | Description                                                   |
| ---------------- | ------------------------------------------------------------- |
| `scoring_method` | Score function used for evaluating structures (e.g., `"bic"`) |
| `start_perm`     | Optional starting permutation of variables                    |
| `use_bes`        | Whether to run optional BES refinement                        |
| `max_iter`       | Maximum number of search iterations                           |
| `random_state`   | Random seed for reproducibility                               |

Note: Starting permutation is optional for the user but is necessary internally since BOSS performs local search over permutations.
You cannot run the algorithm without an initial ordering because the search process improves an existing permutation.

Output:

The learned structure will be stored in: `self.causal_graph_`
Returns: `self.causal_graph_`

Example Usage:
```
from pgmpy.causal_discovery import BOSS

model = BOSS(
    scoring_method="bic",
    random_state=42
)

model.fit(data)

dag = model.causal_graph_
```

Algorithmic Intuition:
BOSS searches over permutations of variables and greedily improves them using a best-move operator. Each permutation is converted to a DAG using Grow-Shrink parent discovery and evaluated using a structure score.

The entire implementation will reside in one file:

`pgmpy/causal_discovery/BOSS.py`

The class will contain the following internal methods:

```
BOSS
 ├── _fit()
 ├── _best_move()
 ├── _score_permutation()
 ├── _project_permutation()
 ├── _grow_shrink()
 ├── _grow()
 ├── _shrink()
 └── _run_bes()
```

@ankurankan please let me know about your thoughts on this design and if any other information is needed

**Comment by anusa-saha** [comment]
Algorithmic Implementation plan

BOSS performs permutation-based causal discovery.

The core idea is:

- Start with a variable permutation
- Convert the permutation into a DAG
- Evaluate the DAG using a structure score
- Apply local permutation moves
- Accept the best scoring move
- Repeat until convergence

Each permutation is converted to a DAG using a Grow-Shrink parent discovery procedure, and evaluated using a scoring function such as BIC.

Optionally, a BES (Backward Equivalence Search) phase can be applied to refine the resulting structure.

**Comment by ankurankan** [comment]
@anusa-saha 

1. I don't think `bes` should be an argument. It should always be run.
2. You might want to inherit from `_ScoreMixin` as well. It already has functionality to compute score delta required for grow-shrink procedure.
3. I don't think you need that many helper methods. Try to think of the minimal linear steps of the algorithm, those would be the main steps that would go into the main `_fit` method. And helper methods should be called within these step. We should avoid having nested called of helper methods.

**Comment by anusa-saha** [comment]
@ankurankan okay thanks for the suggestion will keep these points in my mind while writting the code

**Comment by anusa-saha** [comment]
@ankurankan I have tried the implementation
if i could get an initial review it would be very helpful
